src/Dfc.CourseDirectory.WebV2/Features/Venues/ViewVenues/ViewVenues.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb;
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb.Queries;
+using MediatR;
+
+namespace Dfc.CourseDirectory.WebV2.Features.Venues.ViewVenues
+{
+    public class Query : IRequest<ViewModel>
+    {
+        public Guid ProviderId { get; set; }
+    }
+
+    public class ViewModel
+    {
+        public IReadOnlyCollection<VenueViewModel> Venues { get; set; }
+    }
+
+    public class VenueViewModel
+    {
+        public Guid VenueId { get; set; }
+        public string VenueName { get; set; }
+        public IReadOnlyCollection<string> AddressParts { get; set; }
+        public string PostCode { get; set; }
+    }
+
+    public class Handler : IRequestHandler<Query, ViewModel>
+    {
+        private readonly ICosmosDbQueryDispatcher _cosmosDbQueryDispatcher;
+        private readonly IProviderInfoCache _providerInfoCache;
+
+        public Handler(ICosmosDbQueryDispatcher cosmosDbQueryDispatcher, IProviderInfoCache providerInfoCache)
+        {
+            _cosmosDbQueryDispatcher = cosmosDbQueryDispatcher;
+            _providerInfoCache = providerInfoCache;
+        }
+
+        public async Task<ViewModel> Handle(Query request, CancellationToken cancellationToken)
+        {
+            var provider = await _providerInfoCache.GetProviderInfo(request.ProviderId);
+
+            var results = await _cosmosDbQueryDispatcher.ExecuteQuery(new GetVenuesByProvider { ProviderUkprn = provider.Ukprn });
+
+            return new ViewModel
+            {
+                Venues = results.Select(r => new VenueViewModel
+                {
+                    VenueId = r.Id,
+                    VenueName = r.VenueName,
+                    AddressParts = new[]
+                    {
+                        r.AddressLine1,
+                        r.AddressLine2,
+                        r.Town,
+                        r.County
+                    }.Where(part => !string.IsNullOrWhiteSpace(part)).ToArray(),
```

Comment:
this logic feels like it should be pulled out to something else and unit tested, possibly the cosmos venue model?

Reply:
I diagree - it's ViewModel related logic and we shouldn't be sullying a domain/storage model with it.

Reply:
Because the motivation for this processing isn't clear from the code it's hard for me to judge where the logic to sit. I don't really know *why* this transformation exists. Perhaps you could enlighten me?

Perhaps the viewmodel could expose something specific for the transformation and we could test that?

Reply:
> I don't really know _why_ this transformation exists. Perhaps you could enlighten me?

It's a view model - the transformation exists to prepare the data to be displayed in a view.

> Perhaps the viewmodel could expose something specific for the transformation and we could test that?

What's the difference of doing it here or explicitly on the view model? I don't see the value in unit testing this. Notwithstanding, the output of the view, including this data, is covered by existing tests.

Reply:
> the transformation exists to prepare the data to be displayed in a view

and a layer deeper, why does it need to be displayed this way as opposed to without any transformation at all? Is it a technical consideration (something to do with model shapes, if so why are they different shapes), or is it a business requirement (if so what is that requirement in plain english and where did it come from).

> What's the difference of doing it here or explicitly on the view model

the point for me is to call it out as a required behaviour across the site, intended for reuse wherever we display this data, and a place to document the motivation for the transformation (which remains unclear to me with the current code).

I hope I'm making some sense

Reply:
You're making sense, and I see where you're coming from. However the assumption that all address data throughout the system should be displayed in exactly the same way [hence your desire to standardise it] is perhaps incorrect.

In this particular case, all address fields may not be populated, hence removing the ones that aren't in the view model so as to create less clutter in the view logic. But this logic may not be correct elsewhere on the site, and by creating dependencies across views on formatting logic we introduce a maintenance overhead.

Equally if we ever change the way we want to display addresses, there's a maintenance overhead there too. So basically we're damned if we do, damned if we don't. But in this instance I'm leaning towards maintaining discreet logic for individual views to avoid future headaches and give us maximum flexibility on what is rendered in the view.

Reply:
> perhaps incorrect

I think we should do a quick check with the mighty Westley, he'll probably have an opinion on that one way or the other

---

I think I've exhausted my thoughts on the topic, probably to your relief :grin:

Reply:
We have similar versions of this all over the place:
https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/main/src/Dfc.CourseDirectory.WebV2/SharedViews/Components/ProviderInfoPanelViewComponent.cs#L72-L90
https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/main/src/Dfc.CourseDirectory.WebV2/Features/ApprenticeshipQA/ProviderApprenticeshipQAInfoPanel.cs#L104-L122